build: strip YAML frontmatter from breaking changes in changelog

Planeshifter opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   updates changelog generation to remove YAML blocks from display of breaking changes

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/53.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

Previously, output could include YAML metadata in breaking changes display. Here is an example:

```md
-   [`39e505f`](https://github.com/stdlib-js/stdlib/commit/39e505f50432543c5e6c76c85392e59846ce58c6): remove `math/base/ops/csubf`

    -   To migrate, users should update their require/import paths to use
        `@stdlib/complex/float32/base/sub` which provides the same API and
        implementation.
        Ref: https://github.com/stdlib-js/stdlib/issues/2261
        ---
        type: pre_commit_static_analysis_report
        description: Results of running static analysis checks when committing changes.
        report:
          - task: lint_filenames
            status: passed
          - task: lint_editorconfig
            status: passed
          - task: lint_markdown
            status: na
          - task: lint_package_json
            status: na
          - task: lint_repl_help
            status: na
          - task: lint_javascript_src
            status: na
          - task: lint_javascript_cli
            status: na
          - task: lint_javascript_examples
            status: na
          - task: lint_javascript_tests
            status: na
          - task: lint_javascript_benchmarks
            status: na
          - task: lint_python
            status: na
          - task: lint_r
            status: na
          - task: lint_c_src
            status: na
          - task: lint_c_examples
            status: na
          - task: lint_c_benchmarks
            status: na
          - task: lint_c_tests_fixtures
            status: na
          - task: lint_shell
            status: na
          - task: lint_typescript_declarations
            status: na
          - task: lint_typescript_tests
            status: na
          - task: lint_license_headers
            status: passed
        ---
```

This PR ensures that such blocks are removed.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Planeshifter committed:
build: strip YAML fontmatter from breaking changes in changelog

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/breaking_changes.js:
```diff
@@ -54,9 +87,13 @@ var STDLIB_GITHUB_URL = 'https://github.com/stdlib-js/stdlib/commit';
+	var hash;
+	var out;
+
+	parts = stripYamlBlocks( note.text ).split( '\n' );
```
Seems unnecessary to split, join, split. Why can you not provide a list of lines to `stripYamlBlocks`, return a list, and then perform a trim via a slice by finding the first non-empty string index and the last non-empty string index?

kgryte replied:
Also, trimming seems like something which is outside the scope of stripping YAML blocks. This is no longer relevant provided you refactor to use a slice approach which could happen in `formatBreakingChange`.

Planeshifter committed:
chore: address PR feedback

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/breaking_changes.js:
```diff
@@ -54,13 +86,39 @@ var STDLIB_GITHUB_URL = 'https://github.com/stdlib-js/stdlib/commit';
+
+	for ( i = 0; i < contentLines.length; i++ ) {
+		trimmedLine = trim( contentLines[i] );
+		if ( trimmedLine !== '' ) {
```
What if I have a commit body with empty lines between content?

```text
feat!: foo bar

This commit does X, Y, Z.

BREAKING CHANGE: do beep

To migrate, users should do A, B, C.

The rationale for this change is due to I, J, K.

For greater context, please consult the discussion in issue #123.

---
...YAML
---
```

Planeshifter replied:
@kgryte This function does not operate on commit bodies, but only the parsed "Breaking Change" note.

kgryte replied:
Okay, but still. After a `BREAKING CHANGE` line, why should we compress the content which follows?

kgryte replied:
I've updated the example above accordingly. Previously, the logic would have preserved empty lines. Now those are removed.

Planeshifter replied:
Hmm, yeah, we don't want that ... The trimming here was to ensure that any leading "empty" lines containing only whitespace characters will be removed. 

To implement "via a slice by finding the first non-empty string index and the last non-empty string index", is it safe to assume that we don't have such lines?

kgryte replied:
> is it safe to assume that we don't have such lines?

I am not sure which lines you are asking about. Leading empty lines? Trailing empty lines? Empty lines between paragraphs of content?

kgryte replied:
After further conversation, looks like we can update this logic to simply squeeze multiple empty lines into a single empty line, so that we are not making assumptions regarding the location of YAML blocks in commit message bodies.

Planeshifter replied:
Done.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/breaking_changes.js:
```diff
@@ -54,13 +86,39 @@ var STDLIB_GITHUB_URL = 'https://github.com/stdlib-js/stdlib/commit';
+	hash = trim( note.hash );
+
+	for ( i = 0; i < contentLines.length; i++ ) {
+		trimmedLine = trim( contentLines[i] );
```
This logic isn't the same as trimming the entire note.

Planeshifter committed:
build: refactor to squeeze multiple empty lines into single empty lines

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
chore: clean-up

Signed-off-by: Athan <kgryte@gmail.com>
